Documentation/ServicingGranularity.md

```diff
@@ -0,0 +1,192 @@
+# Servicing Granularity
+
+This doc is a more detailed description of the problems in [source-build#923
+"Figure out how source-build will work with CoreFX per-package servicing
+policy"](https://github.com/dotnet/source-build/issues/923).
```

Comment:
From that issue:

> but we risk shipping a different product than the Microsoft build if we just do that:

Because Microsoft is not using source-build, there is a chance we are not shipping the same thing.
That said, we want source-build to be complete in terms of features, quality and security.
Maybe that is doable by building of the proper branches/tags, and doesn't require granular building.

We need anyhow to be able to source-build from scratch.
So granular building requires extra effort in source-build.

Reply:
> Maybe that is doable by building of the proper branches/tags, and doesn't require granular building.

It is, but I'm trying to describe in this doc what problems this presents. E.g., source-build probably has to build, say, `v3.0.0`, `v3.0.1`, `v3.0.2`, and `v3.0.3` of CoreFX in order to get a prebuiltless SDK with working 3.0.3 runtimes. (Due to needing to be able to build from scratch, as you mentioned.) Then, each of those tags would need to be maintained to bring on new platforms, fix issues, etc., not something Microsoft has been doing.

Could we change the way the repos build to avoid this? (Have the release branches be "[proper](https://github.com/dotnet/source-build/pull/1389#issuecomment-560933567)"?) Maybe--this will need more investigation and work with each repo, and I'm hoping this doc can help motivate investigation like that to find other more reasonable approaches.

I'll add a section about that... trying to get "proper" release branches is ideal, even though I don't know how to get there and it's counter to a longstanding Microsoft servicing routine. I didn't have a proposal at first, but I guess this would be it.

Reply:
> E.g., source-build probably has to build, say, v3.0.0, v3.0.1, v3.0.2, and v3.0.3 of CoreFX in order to get a prebuiltless SDK with working 3.0.3 runtimes. (Due to needing to be able to build from scratch, as you mentioned.)

Maybe we should consider this the driving use-case?

Reply:
It is IMO, I have it in the doc as the top reason I think building every dependency node would be unsustainable: 

> 1. Say the world has changed in a way that makes it impossible to build anything   earlier than CoreFX `v3.0.3` on a new distro or platform. Then, patches need   to be applied to `v3.0.0`, `v3.0.1`, and `v3.0.2` to get them to build,   because the OOB packages are required. The patching work *probably* only   needs to be done once and applied on all affected commits, but as the release   branch continues to be serviced, the amount of work to get the product built   on a new distro balloons.

Did you have somewhere else in mind this drives a specific direction?

Reply:
Sorry for the slow response rate. I've been too-much-multi-tasking.

>  E.g., source-build probably has to build, say, v3.0.0, v3.0.1, v3.0.2, and v3.0.3 of CoreFX in order to get a prebuiltless SDK with working 3.0.3 runtimes

I mean: we should try to make it possible to build .NET Core without having to do builds of multiple versions of a single repository.

I think we are looking at this from the way Microsoft is creating packages for Microsoft build (https://github.com/dotnet/source-build/issues/923).

Maybe we should reverse perspective and look from source-build side:
- What different source packages make sense?
- How can these source packages be assembled from the separate repositories?

> The current workaround: assume tip source code can build old packages

Maybe we can do this without having to call it a workaround?

Reply:
No worries at all, been very busy myself with dotnet/runtime consolidation.

> Maybe we should reverse perspective and look from source-build side:
> 
> * What different source packages make sense?
> * How can these source packages be assembled from the separate repositories?

I think it would be interesting to see if a different arrangement of the package granularity would make more sense for source-build, and ideally have the Microsoft build adopt it. I'm a little skeptical that will provide enough additional value over the existing Microsoft build granularity, though. (In the context of needing value to justify the overhead for .NET Core distro maintainers having to track down and piece together multiple source packages.)

> > The current workaround: assume tip source code can build old packages
> 
> Maybe we can do this without having to call it a workaround?

I'm ok with that if we can get rid of "assume", which to me seems like a pretty big deal. (This is what I believe having "proper" release branches would mean.)

Reply:
In an ideal world anything that ships on its own schedule would have its own repo/branch. There are many reasons why we don't live in that world and probably never will. Unfortunately. On the other end,  re-shipping absolutely everything every time anything in any of the repos changes has real problems too.

I think maybe we can improve "assume" to "validate". In any given month any given package might need to ship, so the code has to be sitting ready to take a  patch. Thus the code in the branch should be trusted as good code matching the last time it shipped until a change is made to ship it again. If a source build distributor would pref to rebuild that code every month, they can do that if they trust that it is good, which we all should based on above reasoning. That does not mean that Microsoft should have to choose to ship the same payload every month to nuget.org for packages that haven't changed.

It terms of validation, I could imagine running tests with the LKG version of a package and the live version always, but not pushing the live version to transport feeds for MS builds or nuget.org, unless there's an explicit opt in. I could further imagine analyzing if source/API has changed, but the "ship it this month" bit isn't flipped, etc.

This is complicated, for sure. I just think that it is not easy to land on either of the simpler solutions: ship the world everywhere always, or separate schedule == separate repo.